### DISCUSSION TITLE
Cannot upload Google Takeout from Desktop app due to CORS issues

### DISCUSSION THREAD

**Discussion by forabi** [ente-io/ente#1854] [Q&A]
### Description

It looks like when uploading a file, the desktop app will receive a redirect to the actual S3/B2 bucket signed URL where the Electron app can directly upload the file to. However, when Electron redirects, it actually drops the `Origin` header (it becomes `null`).

I have tried overriding the request & response headers following [this](https://stackoverflow.com/questions/51254618/how-do-you-handle-cors-in-an-electron-app), but it does not seem like those interceptors can actually access the redirect request at all.

### What product are you using?

Ente Photos

### What platform are you using?

Desktop - Linux

**Comment by mnvr** [comment]
Please refrain from opening issues around self-hosting. It is mentioned in the README.

Re the issue, you should be able to upload takeout if you build and run from the main branch, we already have code that intercepts the headers - https://github.com/ente-io/ente/pull/1753

**Reply by ahxxm** [reply]
Much thanks for the hard work! Is this patch released in Desktop version? I'm using v1.7.4 on Windows, the thumbnails aren't showing up and I can't upload images from Desktop.

Also, regarding limited engineering bandwidth and self-hosting, would it make sense to add a new issue label to group self-hosting concerns and problems? That might also help attract contributors

**Reply by mnvr** [reply]
Yes, the 1.7.4 has that patch. 

> the thumbnails aren't showing up and I can't upload images from Desktop.

In case you haven't seen this already, this might help - https://help.ente.io/self-hosting/guides/configuring-s3

**Reply by ahxxm** [reply]
Oh.. that guide helped me configure for mobile apps, I have been using for a while since and both platform worked fine! 

I'm trying to import from Google Photo takeouts today, the dev console showed a lot CORS errors when 307 redirected to signed urls, like `from origin 'ente://app' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status`

![Untitled-1](https://github.com/user-attachments/assets/c406ffd6-b930-499f-bc61-54df24a27e2c)

I copied a few links and the browser can open them

**Reply by mnvr** [reply]
> `It does not have HTTP ok status`

This sounds like a different issue.

The patch you were referring does a different thing - if the response does not have a "Access-Control-Allow-Origin", or if the "Access-Control-Allow-Origin" was set to "null" in the response, then it replaces it with a "*".

https://github.com/ente-io/ente/blob/968067d6aa344ee59236e22d561c6636765ed0c4/desktop/src/main.ts#L557

The `It does not have HTTP ok status` sounds like the OPTIONS request (the CORS request) to your S3 bucket returned with a non-200 OK response. So you might need to configure something on your S3 bucket, to get it to return 200 OK for these responses. Inspecting the failing requests in the dev tools might give some hints of which extra header or origin (for example) needs to be changed - alternatively you could start with an allow-all CORS and then tighten it to the required ruleset.

(The mobile apps don't use CORS which is why they would've been working even if the S3 bucket did not appropriate CORS allowances)

**Reply by ahxxm** [reply]
TIL mobile apps bypass the check. Updated and verified CORS rule with 

```bash
curl -X OPTIONS https://mybucket.s3.us-west-002.backblazeb2.com/ -H "Origin: ente://app"  -H "Access-Control-Request-Method: GET" -v
< HTTP/1.1 200
< Server: nginx
< access-control-allow-origin: null
< access-control-allow-methods: GET
```

I might understand the issue... The 307 redirections are `GET+preflight`(as I show the METHOD column), and there aren't OPTIONS requests from Desktop app, the redirected request failed preflight checks because it needs a `http ok`, which are `2xx` statuses

**Reply by mnvr** [reply]
You should also ensure CORS on your museum (API server) is setup appropriately to allow the redirect to happen. For debugging, you can start with allowing all origins, headers, methods etc on both the API domain and bucket, and then once you get it working tighten it back appropriately.

**Reply by mnvr** [reply]
(in case you might have some reverse proxies etc in the picture)

**Reply by ahxxm** [reply]
yeah I'm looking at the code, I do have a few reverse proxies(traefik and cloudflare, removed cloudflare but it's the same).

Is the server supposed to return my bucket's domain as well? A quick code search leads me to this [line](https://github.com/ente-io/ente/blob/main/server/cmd/museum/main.go#L956), server returns `ente://app` as is, and dev console also showed that header in the 307 responses

I could patch the code to return `*` but that seems too hacky, I'll try to check how I can pass the bucket's domain to this middleware

**Reply by ahxxm** [reply]
sorry I was totally wrong on "no OPTIONS sent", XHR filter was somehow applied in the dev console, now I can see all the 403 are exactly due to bucket CORS rules, I've applied [this](https://github.com/ente-io/ente/discussions/1764#discussioncomment-9478204) and now it works, and I can upload photos as well.

the only problem is that I also tried to patch server code, now I'm on a newer database schema, I need to stay on local build for while :P
